### PR TITLE
add CSS partial tagged template literal

### LINKED ISSUE #4633: feat: add partial CSS tagged-template literal to fast-element
# 🙋 Feature Request
`@microsoft/fast-element` exposes the `css` tagged-template literal to create CSS documents that support's `ElementStyles`, `CSSDirective` and `string` interpolation. In implementation, each `ElementStyles` instances get's converted to a `CSSStyleSheet` instance, and this works for most purposes.

There are [some cases though](https://github.com/microsoft/fast/pull/4631/files#diff-7cf742765fdd7ef14a385bed41069673a57260992fb0adcd4587cb044e52b316), where a CSS block being interpolated relies on `CSSDirective` interpolation but *does not* result in a full, valid `CSSStyleSheet`.

I want to propose adding a `cssPartial` tagged-template literal that can be interpolated with strings and `CSSDirective` instances, and that returns a `CSSDirective`. This will allow using `CSSDirective`s in re-usable CSS blocks that may not be valid CSS in-and-of themselves, such as property values and collections of rules with no selector:

**Example usage**:
```ts
// 'baseHeightMultiplier', 'density', and 'designUnit' are all 'CSSDirective` instances
const heightNumber: CSSDirective = cssPartial`(${baseHeightMultiplier} + ${density}) * ${designUnit}`;

const myStyles = css`
    host: {
        height: ${heightNumber};
    }
`;
```

### PR DESCRIPTION
**Pull request by nicholasrice**
# Pull Request

## 📖 Description
This PR adds a new CSS tagged template literal called `cssPartial`. `cssPartial` allows abstracting CSS blocks that are not valid CSS in-and-of themselves into a structure usable by `ElementStyles`. `cssPartial` allows interpolation of `CSSDirective` instances as well as strings, and itself returns a `CSSDirective` that can be interpolated into `ElementStyles`.

closes #4633

### General

- [ ] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

### CODE REVIEW

**Review by chrisdholt — APPROVED** [review]

File: packages/web-components/fast-element/docs/guide/leveraging-css.md [hunk 1]
```diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
 You can also pass a CSS `string` or a [CSSStyleSheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet) instance directly to the element definition, or even a mixed array of `string`, `CSSStyleSheet`, or `ElementStyles`.
 :::
 
+### Partial CSS
+There are times when you may want to create reusable blocks of *partial* CSS, where the abstraction is not valid CSS in and of itself, such a groups of CSS properties or a complex value. To do that, you can use the `cssPartial` tagged template literal:
```
Comment (EisenbergEffect) [inline packages/web-components/fast-element/docs/guide/leveraging-css.md:132]:
```suggestion
There are times when you may want to create reusable blocks of *partial* CSS, where the abstraction is not valid CSS in and of itself, such as groups of CSS properties or a complex value. To do that, you can use the `cssPartial` tagged template literal: 
```

**Review by EisenbergEffect — APPROVED** [review]

File: packages/web-components/fast-element/docs/guide/leveraging-css.md (hunk 1 above)
Comment (chrisdholt) [reply packages/web-components/fast-element/docs/guide/leveraging-css.md:132]: nice - my mind walked right past this!

**Comment by EisenbergEffect** [comment]
@nicholasrice One quick thought/question...is there any way to factor out and share code between the `css` and `cssPartial` helpers? If we add bindings in CSS, I'm thinking we'd need to add it in both places as well, so I'm wondering if there's some way to share the basic parsing/collecting code.

**Comment by nicholasrice** [comment]
We can probably share some, though `css` handles a few extra cases which makes re-use a bit tricky.

**Comment by nicholasrice** [comment]
Hmm I just realized `CSSDirective` returns `ComposableStyles` and not just string, so this won't work as-is. Converting to draft

**Review by chrisdholt — APPROVED** [review]

**Comment by nicholasrice** [comment]
@EisenbergEffect Okay I updated this to support all valid return types from `CSSDirective` and also to re-use style collection logic from `css`. 

The main difference between `css` and `cssPartial` is that `css` creates a new CSSStyleSheet for every string in the reduced styles, whereas `cssPartial` will zip-up all strings into a single string and return that from the resulting `CSSDirective`. This means it supports some rather strange syntaxes such as:

**This sets a host's color to blue:**
```ts
class MyDirective extends CSSDirective {
    createCSS() {
        return css`:host { color: blue; }`
    }
}
const partial = cssPartial`color: ${new MyDirective} green`;

css`
    :host {
        ${partial}
    }
`
```

A few questions:
1. Are there general concerns with enabling the above scenario?
2. I currently have `cssPartial` restricted to embedding `string | CSSDirective` (no `ElementStyles or CSSStyleSheet`), but this can obviously be circumvented by returning an `ElementStyles` or `CSSStyleSheet` from a `CSSDirective`. Should I just open up the API?
3. If we changed the behavior of `css` to work like `cssPartial` I think we could technically support both scenarios with one template function, though I think we would loose some of the `CSSStyleSheet` re-use that `ElementStyles` is optimizing for internally. We probably *don't* want to do this, but opening the conversation just in case having one function that does both is preferred.

**Comment by EisenbergEffect** [comment]
Can you expand a bit more on your third point above? I think it would be nice to have a single function, but not sure I fully understand what we'd give up.

**Comment by nicholasrice** [comment]
https://github.com/microsoft/fast/blob/master/packages/web-components/fast-element/src/styles/element-styles.ts#L150 is where ElementStyles converts all style *strings* into CSSStyleSheet instances, caching the instances. That logic is why we can't provide incomplete or invalid CSS to `css` (such as ``` css`color: red` ```). When we try to populate a `CSSStyleSheet` with `"color: red"`, the browser rightfully doesn't know how to convert that to a CSS rule because there is no selector, so it just does nothing.

 If, instead, `ElementStyles` zipped all of the strings into *one* string (or at least all immediate siblings, for instance `[string, string, CSSStyleSheet, string, CSSStyleSheet] => [string, CSSStyleSheet, string, CSSStyleSheet]`), then including a incomplete rule like the above into another `ElementStyles` would just cause the strings to get zipped together prior to `CSSStyleSheet` hydration.

I *think* doing so would reduce the amount of CSSStyleSheet re-use we could do though, because we can no-longer always leverage `CSSStyleSheets` for the individual style strings in a `CSSStyleSheet`.

**Comment by EisenbergEffect** [comment]
Ah, yes. That makes sense. It seems like it could be a perf cost since the browser won't be sharing the sheets or be able to dedupe them internally. So, probably we don't want to do that.

Hmm. I'm not coming up with anything better than the `cssPartial` template function at the moment. If ElementStyles or a CSSStyleSheet are returned, will that break things?

**Comment by nicholasrice** [comment]
Returned from where? `cssPartial`?

**Comment by EisenbergEffect** [comment]
Yes. You mentioned expanding the types to allow that I think. If those were allowed, would it break things?

**Comment by nicholasrice** [comment]
No, it would work the same. It would just make it a little easier to write strange CSS like the above.

**Comment by EisenbergEffect** [comment]
Ok. I think it's ok then. Let's just not put the strange stuff in the docs 😉

**Comment by nicholasrice** [comment]
heh yea okay. I'll update the function.

File: packages/web-components/fast-element/docs/guide/leveraging-css.md [hunk 2]
```diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
...
 :::
 
+### Partial CSS
+There are times when you may want to create reusable blocks of *partial* CSS, where the abstraction is not valid CSS in and of itself, such as groups of CSS properties or a complex value. To do that, you can use the `cssPartial` tagged template literal: 
+
+```ts
+import { css, cssPartial } from "@microsoft/fast-element";
+
+const partial = cssPartial`color: red;`;
+const styles = css`:host{ ${cssPartial} }`;
```
Comment (janechu) [inline packages/web-components/fast-element/docs/guide/leveraging-css.md:138]:
```suggestion
const styles = css`:host{ ${partial} }`;
```

File: packages/web-components/fast-element/docs/guide/leveraging-css.md [hunk 3]
```diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
...
+    }
+  }
+}
+```
+
+### Usage in ElementStyles
+The `CSSDirective` can then be used in an `ElementStyles`, where the CSS string from `createCSS()` will be interpolated into the stylesheet, and the behavior returned from `createBehavior()` will get bound to the element using the stylesheet:
+
+```ts
+const styles = css`:host {${new RandomWidth()}}`;
```
Comment (janechu) [inline packages/web-components/fast-element/docs/guide/leveraging-css.md:193]: Is the intended result that `CSSDirective` class will execute the `crreateCSS()` in the constructor?

**Review by janechu — APPROVED** [review]
Couple clarifying questy

File: packages/web-components/fast-element/docs/guide/leveraging-css.md (hunk 3 above)
Comment (nicholasrice) [reply packages/web-components/fast-element/docs/guide/leveraging-css.md:193]: `createCSS()` gets called by `ElementStyles`, so that happens implicitly

**Pull request merged**